### PR TITLE
Adding subscription support for ethernet interfaces.

### PR DESCRIPTION
**Pull request by nagarwal03**
**What I did:**
Added gNMI subscription transformer support for OpenConfig Ethernet Interfaces.

**Why I did it?**
Replacing existing intf_app.go code with transformer code.

**How did I verify the changes?**
Verified by manually testing gNMI subscription.

**Log files for gNMI subscription manual run:**

[UT logs community subscription ON change IPv4v6.docx](https://github.com/sonic-net/sonic-mgmt-common/files/15395754/UT.logs.community.subscription.ON.change.IPv4v6.docx)
[UT logs Community subscription SAMPLE IPv4v6.docx](https://github.com/sonic-net/sonic-mgmt-common/files/15395755/UT.logs.Community.subscription.SAMPLE.IPv4v6.docx)
[UT logs Community subscription TARGET DEFINED IPv4v6.docx](https://github.com/sonic-net/sonic-mgmt-common/files/15395756/UT.logs.Community.subscription.TARGET.DEFINED.IPv4v6.docx)
[UT logs community subscription - onchange Eth.docx](https://github.com/sonic-net/sonic-mgmt-common/files/15395898/UT.logs.community.subscription.-.onchange.Eth.docx)
[UT logs community subscription TARGET DEFINED ETH.docx](https://github.com/sonic-net/sonic-mgmt-common/files/15396071/UT.logs.community.subscription.TARGET.DEFINED.ETH.docx)
[UT LOGS community subscription SAMPLE Eth.docx](https://github.com/sonic-net/sonic-mgmt-common/files/15396074/UT.LOGS.community.subscription.SAMPLE.Eth.docx)
[UT logs community gnmi subscription TARGET DEFINED intf top level.docx](https://github.com/sonic-net/sonic-mgmt-common/files/15495895/UT.logs.community.gnmi.subscription.TARGET.DEFINED.intf.top.level.docx)
[UT logs community ON CHANGE intf top level.docx](https://github.com/sonic-net/sonic-mgmt-common/files/15495896/UT.logs.community.ON.CHANGE.intf.top.level.docx)
[UT logs community subscription SAMPLE intf top level.docx](https://github.com/sonic-net/sonic-mgmt-common/files/15495897/UT.logs.community.subscription.SAMPLE.intf.top.level.docx)

### CODE REVIEW

**Comment by ranjinidn** [comment]
Ping @sachinholla @anand-kumar-subramanian @mbalachandar for review.

File: translib/transformer/xfmr_intf.go [hunk 1]
```diff
@@ -1361,8 +1447,38 @@ var intf_subintfs_table_xfmr TableXfmrFunc = func(inParams XfmrParams) ([]string
 	var tblList []string
 
 	pathInfo := NewPathInfo(inParams.uri)
+	ifName := pathInfo.Var("name")
 	idx := pathInfo.Var("index")
 
+	if inParams.oper == SUBSCRIBE {
+		var _intfTypeList []E_InterfaceType
+
+		if idx == "*" || idx != "0" {
```
Comment (sachinholla) [inline translib/transformer/xfmr_intf.go:1456]: Shouldn't it be `if idx != "*" && idx != "0" {/* error */}`. Current condition rejects the wildcard key and hence blocks traget_defined subscription at a higher level like interface list

File: translib/transformer/xfmr_intf.go [hunk 2]
```diff
@@ -302,6 +305,30 @@ var intf_table_xfmr TableXfmrFunc = func(inParams XfmrParams) ([]string, error)
 		}
 	}
 
+	if (ifName == "*") && (inParams.oper == SUBSCRIBE) {
+		log.Info("intf_table_xfmr * ifName subscribe with targetUriPath ", targetUriPath)
+
+		if strings.HasPrefix(targetUriPath, "/openconfig-interfaces:interfaces/interface/config") {
+			tblList = append(tblList, "PORT")
+		} else if strings.HasPrefix(targetUriPath, "/openconfig-interfaces:interfaces/interface/state") {
+			tblList = append(tblList, "PORT_TABLE")
```
Comment (sachinholla) [inline translib/transformer/xfmr_intf.go:314]: What about `/interfaces/interface/state/counters` path? It should have been mapped to COUNTERS, right?

File: translib/transformer/xfmr_intf.go [hunk 3]
```diff
@@ -725,7 +752,9 @@ var DbToYang_intf_eth_port_config_xfmr SubTreeXfmrDbToYang = func(inParams XfmrP
 	intfsObj := getIntfsRoot(inParams.ygRoot)
 	pathInfo := NewPathInfo(inParams.uri)
 	uriIfName := pathInfo.Var("name")
-	ifName := uriIfName
+	ifName := *(&uriIfName)
```
Comment (sachinholla) [inline translib/transformer/xfmr_intf.go:755]: What does this do? How it is different from normal assignment??

File: translib/transformer/xfmr_intf.go [hunk 4]
```diff
@@ -725,7 +752,9 @@ var DbToYang_intf_eth_port_config_xfmr SubTreeXfmrDbToYang = func(inParams XfmrP
...
+
+	log.Infof("DbToYang_intf_eth_port_config_xfmr: Interface name : %s ", ifName)
```
Comment (sachinholla) [inline translib/transformer/xfmr_intf.go:757]: This can be a verbose log

File: translib/transformer/xfmr_intf.go [hunk 5]
```diff
@@ -801,6 +830,63 @@ var DbToYang_intf_eth_port_config_xfmr SubTreeXfmrDbToYang = func(inParams XfmrP
 	return err
 }
 
+var Subscribe_intf_eth_port_config_xfmr SubTreeXfmrSubscribe = func(inParams XfmrSubscInParams) (XfmrSubscOutParams, error) {
+	var err error
+	var result XfmrSubscOutParams
+
+	if inParams.subscProc == TRANSLATE_SUBSCRIBE {
+		log.Info("Subscribe_intf_eth_port_config_xfmr: inParams.subscProc: ", inParams.subscProc)
```
Comment (sachinholla) [inline translib/transformer/xfmr_intf.go:838]: I feel this log is unnecessary.. You have already checked for the `inParams.subscProc` value and taken a decision. Make it a verbose log if you want to retain

File: translib/transformer/xfmr_intf.go [hunk 6]
```diff
@@ -801,6 +830,63 @@ var DbToYang_intf_eth_port_config_xfmr SubTreeXfmrDbToYang = func(inParams XfmrP
...
+
+		pathInfo := NewPathInfo(inParams.uri)
+		targetUriPath := pathInfo.YangPath
+
+		log.Infof("Subscribe_intf_eth_port_config_xfmr:- URI:%s pathinfo:%s ", inParams.uri, pathInfo.Path)
+		log.Infof("Subscribe_intf_eth_port_config_xfmr:- Target URI path:%s", targetUriPath)
```
Comment (sachinholla) [inline translib/transformer/xfmr_intf.go:844]: These logs are unnecessary as these params (uri, targeturi etc) are not used in this code path. Change to verbose log if you want to retain

File: translib/transformer/xfmr_intf.go [hunk 7]
```diff
@@ -801,6 +830,63 @@ var DbToYang_intf_eth_port_config_xfmr SubTreeXfmrDbToYang = func(inParams XfmrP
...
+		result.nOpts.mInterval = 15
+		result.isVirtualTbl = false
+		result.needCache = true
+
+		ifName := pathInfo.Var("name")
+		log.Info("Subscribe_intf_eth_port_config_xfmr: ifName: ", ifName)
+
+		if ifName == "" {
+			ifName = "*"
+		}
```
Comment (sachinholla) [inline translib/transformer/xfmr_intf.go:858]: Shorthand: `ifName := pathInfo.StringVar("name", "*")`

File: translib/transformer/xfmr_intf.go [hunk 8]
```diff
@@ -801,6 +830,63 @@ var DbToYang_intf_eth_port_config_xfmr SubTreeXfmrDbToYang = func(inParams XfmrP
...
+
+		result.dbDataMap = RedisDbSubscribeMap{db.ConfigDB: {
+			"PORT": {ifName: {"autoneg": "auto-negotiate", "speed": "port-speed"}}}}
+
+		log.Info("Subscribe_intf_eth_port_config_xfmr: result ", result)
```
Comment (sachinholla) [inline translib/transformer/xfmr_intf.go:863]: Change to verbose log please

File: translib/transformer/xfmr_intf.go [hunk 9]
```diff
@@ -801,6 +830,63 @@ var DbToYang_intf_eth_port_config_xfmr SubTreeXfmrDbToYang = func(inParams XfmrP
...
+	}
+	return result, err
+}
+
+var DbToYangPath_intf_eth_port_config_path_xfmr PathXfmrDbToYangFunc = func(params XfmrDbToYgPathParams) error {
+	log.Info("DbToYangPath_intf_eth_port_config_path_xfmr: params: ", params)
```
Comment (sachinholla) [inline translib/transformer/xfmr_intf.go:869]: User verbose log while logging the whole struct..

File: translib/transformer/xfmr_intf.go [hunk 10]
```diff
@@ -801,6 +830,63 @@ var DbToYang_intf_eth_port_config_xfmr SubTreeXfmrDbToYang = func(inParams XfmrP
...
+
+	intfRoot := "/openconfig-interfaces:interfaces/interface"
+
+	if params.tblName != "PORT" {
+		log.Info("DbToYangPath_intf_eth_port_config_path_xfmr: from wrong table: ", params.tblName)
+		return nil
+	}
+
+	if (params.tblName == "PORT") && (len(params.tblKeyComp) > 0) {
+		params.ygPathKeys[intfRoot+"/name"] = *(&params.tblKeyComp[0])
```
Comment (sachinholla) [inline translib/transformer/xfmr_intf.go:879]: Why `*(&xyz)` do?

File: translib/transformer/xfmr_intf.go [hunk 11]
```diff
@@ -2452,17 +2562,177 @@ var Subscribe_intf_ip_addr_xfmr = func(inParams XfmrSubscInParams) (XfmrSubscOut
...
+		uriIfName := pathInfo.Var("name")
+		tableName := ""
+		ipKey := ""
+		ifKey := ""
+
+		if uriIfName == "" || uriIfName == "*" {
+			ifKey = "*"
+		} else {
+			sonicIfName := &uriIfName
+			ifKey = *sonicIfName
```
Comment (sachinholla) [inline translib/transformer/xfmr_intf.go:2615]: Why take an address and then dereference? I think just the following is sufficient -- `ifKey := pathInfo.StringVar("name", "*")`

File: translib/transformer/xfmr_intf.go (hunk 2 above)
Comment (nagarwal03) [reply translib/transformer/xfmr_intf.go:314]: The Subscribe_intf_get_counters_xfmr handles the mapping of interfaces/interface/state/counters to COUNTERS_PORT_NAME_MAP

File: translib/transformer/xfmr_intf.go (hunk 1 above)
Comment (nagarwal03) [reply translib/transformer/xfmr_intf.go]: Thanks for reviewing the PR @sachinholla. I have addressed the comments. Please review.

**Review by mbalachandar — APPROVED** [review]

**Review by sachinholla — APPROVED** [review]
Changes look fine.. Please verify target_defined subscription for `/interfaces/interface[name=*]` path as well.

**Pull request merged**